app/src/main/java/com/hedvig/app/feature/embark/util/VariableExtractor.kt

```diff
@@ -0,0 +1,123 @@
+package com.hedvig.app.feature.embark.util
+
+import com.hedvig.android.owldroid.fragment.GraphQLVariablesFragment
+import com.hedvig.android.owldroid.type.EmbarkAPIGraphQLSingleVariableCasting
+import com.hedvig.android.owldroid.type.EmbarkAPIGraphQLVariableGeneratedType
+import com.hedvig.app.feature.embark.ValueStore
+import com.hedvig.app.util.plus
+import com.hedvig.app.util.toJsonObject
+import java.util.UUID
+import org.json.JSONObject
+
+object VariableExtractor {
+
+    fun extractVariables(variables: List<GraphQLVariablesFragment>, valueStore: ValueStore): JSONObject {
+        var regularVariables = variables.mapNotNull {
+            extractRegularVariables(it, valueStore)
+        }.toJsonObject()
+
+        val multiActionVariables = variables.mapNotNull {
+            extractMultiActionVariables(it, valueStore)
+        }
+
+        multiActionVariables.forEach {
+            regularVariables = regularVariables.plus(it)
+        }
+        return regularVariables
+    }
+
+    private fun extractRegularVariables(
+        variable: GraphQLVariablesFragment,
+        valueStore: ValueStore
+    ): Pair<String, Any>? {
+        variable.asEmbarkAPIGraphQLSingleVariable?.let { singleVariable ->
+            val storeValue = valueStore.get(singleVariable.from)
+            return storeValue?.let {
+                createSingleVariable(singleVariable.key, it, singleVariable.as_)
+            }
+        }
+        variable.asEmbarkAPIGraphQLGeneratedVariable?.let { generatedVariable ->
+            val generatedValue = UUID.randomUUID()
+            valueStore.put(generatedVariable.storeAs, generatedValue.toString())
+            return createGeneratedVariable(
+                generatedVariable.key,
+                generatedVariable.type,
+                generatedValue
+            )
+        }
+        return null
+    }
+
+    private fun extractMultiActionVariables(
+        variable: GraphQLVariablesFragment,
+        valueStore: ValueStore
+    ): JSONObject? {
+        variable.asEmbarkAPIGraphQLMultiActionVariable?.let { multiActionVariables ->
+            val multiActionItems = valueStore.getMultiActionItems(multiActionVariables.key)
+            val casted = multiActionItems.map { multiActionMap ->
```

Comment:
If we squint a little at this function, it looks like it does the same job as `extractVariables` does at top-level, but with a `ValueStore` scoped to the Multi Action-item. Perhaps we could make that the reality in code, as well? Then we would not have to express the procedure twice, so to speak.

We could accomplish this by translating the `GraphQLVariablesFragment.AsGraphQLMultiActionVariable` into a `GraphQLVariablesFragment`, creating a static `ValueStore` from the `multiActionItems`, and then finally calling `extractVariables` with those two parameters as the argument.

What are your thoughts on that approach?

Reply:
Yes I was thinking the same, but had difficulty creating the fragment in `embarkStory.graphql` - how would that look? Thanks